src/ethereum/prague/vm/eof/validation.py

```diff
@@ -0,0 +1,396 @@
+    ----------
+    validator : Validator
+        The validator for the EOF container.
+
+    Raises
+    ------
+    InvalidEof
+        If the code section is invalid.
+    """
+    eof_meta = validator.eof.metadata
+    validator.referenced_subcontainers = {
+        Ops.EOFCREATE: [],
+        Ops.RETURNCONTRACT: [],
+    }
+    for code_index, code in enumerate(eof_meta.code_section_contents):
+        validator.current_index = Uint(code_index)
+        validator.current_code = code
+        validator.current_pc = Uint(0)
+        validator.sections[validator.current_index] = {}
+        validate_code_section(validator)
+
+    if validator.has_return_contract and (
+        validator.has_return or validator.has_stop
+    ):
+        raise InvalidEof("Container has both RETURNCONTRACT and STOP/RETURN")
+
+    if eof_meta.num_container_sections > 0:
+        eofcreate_references = validator.referenced_subcontainers[
+            Ops.EOFCREATE
+        ]
+        returncontract_references = validator.referenced_subcontainers[
+            Ops.RETURNCONTRACT
+        ]
+        for index in range(len(eof_meta.container_section_contents)):
+            if (
+                index in eofcreate_references
+                and index in returncontract_references
+            ):
+                raise InvalidEof(
+                    "Container referenced by both EOFCREATE and RETURNCONTRACT"
+                )
+            elif (
+                index not in eofcreate_references
+                and index not in returncontract_references
+            ):
+                raise InvalidEof("Container never referenced")
+
+            sub_validator = validate_eof_container(
+                eof_meta.container_section_contents[index], False
+            )
+            if index in eofcreate_references and (
+                sub_validator.has_stop or sub_validator.has_return
+            ):
+                raise InvalidEof(
+                    "Container referenced by EOFCREATE has STOP or RETURN"
+                )
+            if (
+                index in returncontract_references
+                and sub_validator.has_return_contract
+            ):
+                raise InvalidEof(
+                    "Container referenced by RETURNCONTRACT has RETURNCONTRACT"
+                )
+
+
+def validate_eof_container(
+    container: bytes, is_init_container: bool
+) -> Validator:
+    """
+    Validate the Ethereum Object Format (EOF) container.
+
+    Parameters
+    ----------
+    container : bytes
+        The EOF container to validate.
+    is_init_container : bool
+        Whether the container is an init container for EOFCREATE/
+        create transactions.
+
+    Raises
+    ------
+    InvalidEof
+        If the EOF container is invalid.
+    """
+    # Validate the magic
+    if (
+        len(container) < EOF_MAGIC_LENGTH
+        or container[:EOF_MAGIC_LENGTH] != EOF_MAGIC
+    ):
+        raise InvalidEof("Invalid magic")
```

Comment:
Similarly, maybe:

```suggestion
    if len(container) < EOF_MAGIC_LENGTH:
        raise InvalidEof("Too short for magic")
    if container[:EOF_MAGIC_LENGTH] != EOF_MAGIC:
        raise InvalidEof("Invalid magic")
```

Personally I find separate `if` blocks easier to read than a single long `if` condition.